src/backend/utils/resgroup/resgroup.c

```diff
@@ -4766,6 +4809,49 @@ moveQueryCheck(int sessionId, Oid groupId)
 	cdbdisp_clearCdbPgResults(&cdb_pgresults);
 }
 
+/*
+ * Try to give away all slot control to target process.
+ */
+static void
+resGroupGiveSlotAway(int sessionId, void *slot, Oid groupId)
+{
+	long timeout;
+	int64 curTime;
+	int64 waitStart;
+	int latchRes;
+	
+	SIMPLE_FAULT_INJECTOR("resource_group_give_away1");
+
+	if (!ResGroupMoveSignalTarget(sessionId, slot, InvalidOid))
+		elog(ERROR, "can't send signal to process");
+	
+	waitStart = GetCurrentIntegerTimestamp();
+
+	SIMPLE_FAULT_INJECTOR("resource_group_give_away2");
+
+	for (;;)
+	{
+		ResetLatch(&MyProc->procLatch);
+
+		CHECK_FOR_INTERRUPTS();
+
+		curTime = GetCurrentIntegerTimestamp();
+		timeout = gp_resource_group_move_timeout - (curTime - waitStart) / 1000;
+		if (timeout < 0)
+			break;
+
+		latchRes = WaitLatch(&MyProc->procLatch,
+			WL_LATCH_SET | WL_TIMEOUT | WL_POSTMASTER_DEATH, timeout);
+
+		if (latchRes & WL_TIMEOUT)
+			break;
+		if (latchRes & WL_POSTMASTER_DEATH)
+			elog(ERROR, "got WL_POSTMASTER_DEATH waiting on latch; exiting...");
+	}
```

Comment:
Do we go to the next iteration in the case of WL_LATCH_SET?

We need to check our condition (remote process handle signal) before leaving this loop. This latch may be set in many other places for different purposes (e.g. in SIGHUP [handler](https://github.com/arenadata/gpdb/blob/d4534e08c27ede9b572ecfcf0f167518bf562ff4/src/backend/tcop/postgres.c#L3614) to reread config). And, as result, we may leave this loop before time is running up and the target process handles the signal.